### PR TITLE
Support for JetStream-based lattice cache

### PR DESCRIPTION
**Pull request by autodidaddict**
Closing #136 #153

### CODE REVIEW

**Comment by autodidaddict** [comment]
I know the tests aren't going to work yet - I need to figure out how to run a jetstream-enabled version of nats server in the CI pipeline.

**Comment by autodidaddict** [comment]
w00t! CI tests pass with JetStream.

File: host_core/lib/host_core/application.ex [hunk 1]
```diff
@@ -62,72 +70,41 @@ defmodule HostCore.Application do
...
-             %{topic: "wasmbus.rpc.#{config.lattice_prefix}.refmaps.put"},
+               topic: "wasmbus.ctl.#{config.lattice_prefix}.linkdefs.*",
+               queue_group: "wasmbus.ctl.#{config.lattice_prefix}"
+             },
              %{
-               topic: "wasmbus.rpc.#{config.lattice_prefix}.refmaps.get",
-               queue_group: "wasmbus.rpc.#{config.lattice_prefix}.refmaps.get"
-             }
+               topic: "wasmbus.ctl.#{config.lattice_prefix}.get.>",
+               queue_group: "wasmbus.ctl.#{config.lattice_prefix}"
```
Comment (brooksmtownsend) [inline host_core/lib/host_core/application.ex:88]: Are there any implications here for `linkdef` operations and `get` operations to operate on the same queue group?

File: host_core/lib/host_core/control_interface/server.ex [hunk 2]
```diff
@@ -68,129 +75,138 @@ defmodule HostCore.ControlInterface.Server do
...
+  defp handle_request({"linkdefs", "del"}, body, _reply_to) do
+    ld = Jason.decode!(body)
+
+    HostCore.Linkdefs.Manager.del_link_definition(
+      ld["actor_id"],
+      ld["contract_id"],
+      ld["link_name"]
+    )
+
+    {:reply, ""}
```
Comment (brooksmtownsend) [inline host_core/lib/host_core/control_interface/server.ex:112]: For our other operations we've been following a pattern of responding with a reply object, usually just indicating if the operation was successful and if not then a failure message. Can we indicate the status of the del_link / put_link in that reply?

For strongly typed languages like Rust, we may want to consider standardizing the `ack` that we do respond with, e.g. just an object with success and error message

File: host_core/lib/host_core/control_interface/server.ex [hunk 3]
```diff
@@ -68,129 +75,138 @@ defmodule HostCore.ControlInterface.Server do
...
-        end
+    response =
+      case HostCore.Actors.ActorSupervisor.live_update(update_actor_command["new_actor_ref"]) do
+        :ok -> Jason.encode!(%{accepted: true})
+        :error -> Jason.encode!(%{accepted: false})
+      end
 
-      {:reply, response}
-    end
+    {:reply, response}
```
Comment (brooksmtownsend) [inline host_core/lib/host_core/control_interface/server.ex:203]: re: above comment, it really seems like we vary from `""` to `%{}` to a full response object, I propose that all of our responses are the same JSON shape and just indicate success and provide an error message if applicable. Thoughts?

File: host_core/lib/host_core/jetstream/cache_loader.ex [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    |> String.split(".")
+    # lc
+    |> List.delete_at(0)
+    # {prefix}
+    |> List.delete_at(0)
+    # {type, key}
+    |> List.to_tuple()
+    |> handle_request(body)
+
+    {:reply, ""}
```
Comment (brooksmtownsend) [inline host_core/lib/host_core/jetstream/cache_loader.ex:16]: Same comment re: standard reply ack object

File: host_core/lib/host_core/jetstream/client.ex [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+      deliver_subject: config.cache_deliver_inbox
+    }
+
+    {:ok, state, {:continue, :ensure_stream}}
+  end
+
+  @impl true
+  def handle_continue(:ensure_stream, state) do
+    # TODO get rid of this
+    Process.sleep(100)
```
Comment (brooksmtownsend) [inline host_core/lib/host_core/jetstream/client.ex:22]: Are you able to get rid of this 🙂

File: host_core/lib/host_core/jetstream/client.ex [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      %{
+        stream_name: stream_name,
+        name: consumer_name,
+        config: %{
+          description: "cache loader for #{state.lattice_prefix}",
+          ack_policy: "explicit",
+          filter_subject: ">",
+          deliver_policy: "last_per_subject",
+          deliver_subject: state.deliver_subject,
+          # idle_heartbeat: 2_000_000_000,
```
Comment (brooksmtownsend) [inline host_core/lib/host_core/jetstream/client.ex:81]: Do we want to set this to something like `30_000`, or just remove it?

**Review by brooksmtownsend — CHANGES_REQUESTED** [review]
A couple suggestions, mostly involving a standard way to respond to control interface queries

File: host_core/lib/host_core/application.ex (hunk 1 above)
Comment (autodidaddict) [reply host_core/lib/host_core/application.ex:88]: the queue group defines the "psuedo random selection" list from among potential subscribers to a given topic name. By setting the queue group name to what it is, we're basically just guaranteeing that for each topic where appropriate, the subscribers are chosen randomly from the list of all subscribers in the lattice. 

tl;dr this should be fine.

File: host_core/lib/host_core/jetstream/client.ex (hunk 5 above)
Comment (autodidaddict) [reply host_core/lib/host_core/jetstream/client.ex:22]: Not yet. We need the other consumer servers running before we can fire up the jetstream client commands.

File: host_core/lib/host_core/jetstream/client.ex (hunk 6 above)
Comment (autodidaddict) [reply host_core/lib/host_core/jetstream/client.ex:81]: idle heartbeat needs to be non-existent at this point. This controls how often the NATS server sends _us_ heartbeat messages, which we don't support yet (but we will when the `Gnat` client gets JetStream support)

File: host_core/lib/host_core/application.ex (hunk 1 above)
Comment (brooksmtownsend) [reply host_core/lib/host_core/application.ex:88]: Cool that's what I figured 👍🏻

**Review by brooksmtownsend — APPROVED** [review]

**Comment by brooksmtownsend** [comment]
@autodidaddict BTW the checks won't run until the merge conflicts are addressed

**Comment by autodidaddict** [comment]
Resolve merge conflicts? Pffft. YOLO.

**Review by brooksmtownsend — APPROVED** [review]
